server/src/main/java/org/apache/cassandra/sidecar/livemigration/CassandraInstanceFilesImpl.java

```diff
@@ -0,0 +1,145 @@
+    {
+        this.instanceMetadata = instanceMetadata;
+        this.configuration = configuration;
+    }
+
+    @Override
+    public List<InstanceFileInfo> getFiles() throws IOException
+    {
+        return getFiles(configuration.filesToExclude(), configuration.directoriesToExclude());
+    }
+
+    private List<InstanceFileInfo> getFiles(Set<String> filesToExclude,
+                                            Set<String> dirsToExclude) throws IOException
+    {
+        List<DirVisitor> dirVisitors = getDirVisitorList(filesToExclude, dirsToExclude);
+        List<InstanceFileInfo> instanceFileInfos = new ArrayList<>();
+
+        for (DirVisitor dirVisitor : dirVisitors)
+        {
+            instanceFileInfos.addAll(dirVisitor.getFiles());
+        }
+
+        LOGGER.info("{} number of files are getting listed", instanceFileInfos.size());
+        return instanceFileInfos;
+    }
+
+    public List<DirVisitor> getDirVisitorList(Set<String> filesToExclude,
+                                              Set<String> dirsToExclude)
+    {
+        List<DirVisitor> dataFilesToVisit = new ArrayList<>();
+        List<String> dirsToCopy = LiveMigrationInstanceMetadataUtil.dirsToCopy(instanceMetadata);
+        Map<String, String> dirPathPrefix = LiveMigrationInstanceMetadataUtil.dirPathPrefixMap(instanceMetadata);
+        Map<String, Set<String>> dirPlaceholderMap = LiveMigrationInstanceMetadataUtil.dirPlaceHoldersMap(instanceMetadata);
+
+        for (String dir : dirsToCopy)
+        {
+            getDirToVisit(dir,
+                          dirPlaceholderMap.get(dir),
+                          dirPathPrefix.get(dir),
+                          filesToExclude,
+                          dirsToExclude)
+            .ifPresent(dataFilesToVisit::add);
+        }
+
+        return dataFilesToVisit;
+    }
+
+    Optional<DirVisitor> getDirToVisit(String homeDir, Set<String> placeholders, String pathPrefix,
+                                       Set<String> filesToExclude, Set<String> dirsToExclude)
+    {
+        if (null == homeDir)
+        {
+            return Optional.empty();
+        }
+        Path homeDirPath = Paths.get(homeDir);
+        if (!Files.exists(homeDirPath))
+        {
+            LOGGER.info("Directory {} does not exist.", homeDir);
+            return Optional.empty();
+        }
+
+        Objects.requireNonNull(placeholders, "placeholders are required");
+        Objects.requireNonNull(pathPrefix, "pathPrefix is required");
+
+        Set<PathMatcher> fileExclusionMatchers = toPathMatchers(filesToExclude, placeholders, homeDir);
+        Set<PathMatcher> dirExclusionMatchers = toPathMatchers(dirsToExclude, placeholders, homeDir);
+
+        return Optional.of(new DirVisitor(homeDir, pathPrefix, fileExclusionMatchers, dirExclusionMatchers));
+    }
+
+    Set<PathMatcher> toPathMatchers(Set<String> exclusions, Set<String> homeDirPlaceholders, String homeDir)
+    {
+        if (null == exclusions)
```

Comment:
nit: check empty case too. 

```suggestion
        if (null == exclusions || exclusions.isEmpty())
```

Reply:
done